src/coreclr/jit/optimizer.cpp

```diff
@@ -4570,6 +4570,344 @@ PhaseStatus Compiler::optUnrollLoops()
+
+        // Check that we have linear flow and are still in the same EH region
+
+        if (middleBlock->GetUniquePred(this) == nullptr)
+        {
+            return false;
+        }
+
+        if (!BasicBlock::sameEHRegion(middleBlock, block))
+        {
+            return false;
+        }
+
+        // Can all the nodes within the middle block be made to conditionally execute?
+        for (Statement* const stmt : middleBlock->Statements())
+        {
+            GenTree* tree = stmt->GetRootNode();
+            switch (tree->gtOper)
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->gtGetOp1();
+                    GenTree* op2 = tree->gtGetOp2();
+
+                    // Only one per assignment per block can be conditionally executed.
+                    if (asgNode != nullptr || op2->OperIs(GT_SELECT))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the destination of the assign is a local variable with integer type.
+                    if (!op1->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op1))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the nodes of the assign won't cause any additional side effects.
+                    if ((op1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+                        (op2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+
+                    asgNode = tree;
+                    asgStmt = stmt;
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    if (tree->gtGetOp1() != nullptr || (tree->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+    }
+    if (asgNode == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("\nConditionally executing " FMT_BB " inside " FMT_BB "\n", middleBlock->bbNum, block->bbNum);
+        for (BasicBlock* dumpBlock = block; dumpBlock != middleBlock->bbNext; dumpBlock = dumpBlock->bbNext)
+        {
+            fgDumpBlock(dumpBlock);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    // Using SELECT nodes means that full assignment is always evaluated.
+    // Put a limit on the original source and destination of the assignment.
+    if (!compStressCompile(STRESS_IF_CONVERSION_COST, 25))
+    {
+        int cost = asgNode->gtGetOp2()->GetCostEx() + (gtIsLikelyRegVar(asgNode->gtGetOp1()) ? 0 : 2);
```

Comment:
```suggestion
        int cost = asgNode->gtGetOp2()->GetCostEx();
        if (!gtIsLikelyRegVar(asgNode->gtGetOp1())
        {
            cost += asgNode->gtGetOp1()->GetCostEx();
        }
```

Sorry, I realized that we can still have arbitrary nodes on the LHS that we should take into account.

Reply:
..nevermind, it was checked to be a local above.